### DISCUSSION TITLE
Now dev server seems to need a restart every X hot reloads

### DISCUSSION THREAD

**Discussion by aloukissas** [vercel/vercel#4143] [General]
This looks like a memory leak of some sort. After a while of editing -> saving -> hot reload, reloads start to get slower up to the point that they don't happen. Killing the dev server and restarting it fixes the problem. I'm running `now dev` on version 18.0.

**Comment by coetry** [comment]
Hi @aloukissas can you let us know if you're experiencing this with the latest canary? 

`npm i -g now@canary` 

Also, which framework were you using with `now dev`?

**Reply by aloukissas** [reply]
I'll try the newest version. This is with nextjs.

**Reply by aloukissas** [reply]
@coetry no improvement really. It gets progressively worse until it craps out. Smells like a memleak.

**Reply by roelvan** [reply]
I am having the same issue with Next.js as you describe.

**Reply by aloukissas** [reply]
Continuing to have an issue with the 19.x canary version, still :/

**Reply by ctrlplusb** [reply]
I'm experiencing the same. Pretty killer on the dev experience. Running a Create React App frontend and a single lambda. Get through a handful of edits on the CRA until it won't load the UI anymore. Stopping the `dev` command also leaves a zombie process in play locking the 3000 port. I have to run `killall -9 node` to clean up correctly before restarting.

**Comment by styfle** [comment]
Do you have the same issue with `next dev`?

You can add this to your package.json:

```json
{
  "scripts": {
    "dev": "next dev"
  }
}
```

Then run `yarn dev` or `npm run dev` and see if the issue happens again.

**Comment by ctrlplusb** [comment]
For the sake of my comment not being buried in a reply thread...

> I'm experiencing the same. Pretty killer on the dev experience. Running a Create React App frontend and a single lambda. Get through a handful of edits on the CRA until it won't load the UI anymore. Stopping the `dev` command also leaves a zombie process in play locking the 3000 port. I have to run `killall -9 node` to clean up correctly before restarting.

**Reply by ctrlplusb** [reply]
For now I have resorted to running the Create React App dev server directly:

```bash
PORT=4000 yarn create-react-app start
```

I added a proxy config to the `package.json` so that I am still able to access my serverless functions:

```
"proxy": "http://localhost:3000"
```

I also updated the `develop` command for `vercel dev` to perform simple echo out to the console. So essentially it doesn't run and proxy the CRA server instance, but it still builds and runs the serverless functions.

Now if I have to do a restart it's a lot faster. Still not ideal, but I can work with this for now.

Also, make sure you run the following command to identify any zombie process left by `vercel dev` on port 3000, otherwise the `vercel dev` command will bind a new process on the next port up (i.e. 3001, 3002, etc)

```bash
sudo lsof -i :3000 
```

If you find a process you'll need to pass the process id to the kill command:

```bash
kill -9 <PID>
```

> FYI, I am using OSX, so these commands may not work for everyone else. 🤗

**Reply by ctrlplusb** [reply]
@coetry - FYI. Some insight into the pain being experienced here.

I'm surprised there hasn't been more kickback in regards to this.

**Reply by ctrlplusb** [reply]
Also worth noting that I am not even executing any UI framework via `vercel dev`. Just the reloading of my single GraphQL API endpoint a few times is enough to cause the memory leak / zombie process.

**Comment by ctrlplusb** [comment]
I wonder if [this](https://github.com/zeit/now/issues/3451#issuecomment-618510917) is the cause of the memory leak?

**Comment by ctrlplusb** [comment]
Ok so the development experience was driving me close to insanity. I decided to give a go at writing my own development script. I've been running it now for 2 hours straight with 0 restarts required. 

Some rad features:

- Super fast. I don't do a full TypeScript / Webpack recompilation of my API endpoints for every code change. A single TS compilation instance is shared between restarts. What a dream.
- Full hot reloading - even after I install new dependencies or regenerate my GraphQL schema etc. Plus I hooked up `react-refresh` for an even more blazing experience in developing the UI.

Suffice to say I am much more stoked right now.  When I finish some current commitments I'll wrap this into a little library that others can use until the `vercel dev` experience improves.

Early warning though - there will be a few limitations:

- It won't show the typical 404/500 error pages that are typically served by the Vercel development server.
- It will only work for `node` based APIs. I have no interest in supporting all the rest.
- It doesn't support any complex function pathing configurations via the vercel json. It does straight up mapping against the structure within your `api` directory.

Those limitations don't bother myself at all, but I'll be sure to make the library open source to allow for forking / PRs etc.

I'll be sure to ping you @aloukissas 🙏

---

_Update:_ 8 hours now, no restarts, fully active dev. A full workday. 🎉

That's a wrap. 💜

**Reply by EvanLovely** [reply]
Ohhh!! Would love to get a glimpse at this! Even un-polished code custom to your setup would be OK by as I can parse it out! Thanks!

**Reply by ctrlplusb** [reply]
Not announcing this quite yet as still wanna use it for a bit longer to make sure DX is as rad as it can be. It's working pretty well though and I even added support for custom routing config and other core features of Vercel.

https://github.com/ctrlplusb/vercel-node-dev

Would be great to get your feedback. 

As stated in the docs, this is for any Node.js runtime project that isn't Next.js. :)  Next.js has it's own rad dev server.

**Reply by ctrlplusb** [reply]
@aloukissas @EvanLovely

**Comment by mcsdevv** [comment]
Hey everyone, yesterday we released 19.1.0 with a number of upgrades. Could I ask you all to give it a try and let us know if this helps?
